### PR TITLE
feat(core): add boxplot chart

### LINKED ISSUE #850: Design specs for Boxplot chart
Issue description:

Doing design specs explorations (box, minimum, maximum, median, whiskers, tooltip treatment, etc.) for boxplot chart.

### PR DESCRIPTION
**Pull request by theiliad**
image

**TODO:**

- [x] add proper event dispatching for box elements
- [x] add re-render logic
- [x] horizontal & vertical orientations

Closes #850

### CODE REVIEW

**Comment by vgb1993** [comment]
Hello guys, since you are actively developing this feature I was wondering if will support:
 + Multiple datasets. 
 + Horizontal orientation.
 + Time series.

I'd like to use this chart to display scheduled shipments for each driver.

Similar to a gantt chart: 
https://chandoo.org/img/pm/gantt-box-chart-template-download.png
https://github.com/scholtzan/d3-gantt.

Cheers!

**Comment by theiliad** [comment]
> Hello guys, since you are actively developing this feature I was wondering if will support:
> 
> * Multiple datasets.
> * Horizontal orientation.
> * Time series.
> 
> I'd like to use this chart to display scheduled shipments for each driver.
> 
> Similar to a gantt chart:
> https://chandoo.org/img/pm/gantt-box-chart-template-download.png
> https://github.com/scholtzan/d3-gantt.
> 
> Cheers!

Hi, I think there are fundamental differences between what you've linked in your comments and what this PR is adding. Boxplots are usually used to show the distribution of elements in the data, what you're showing is basically a gantt chart

**Comment by vgb1993** [comment]
Yes there are diferences, gantt is a chart of its own. Still, a horizontal boxplot looks kind of simillar.

Anyway, cheers for adding horizontal support!

File: packages/core/src/components/graphs/boxplot.ts [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+			(d, i?) => this.services.cartesianScales.getDomainValue(d, i),
+			(d, i?) => this.services.cartesianScales.getRangeValue(d, i),
+			orientation
+		);
+
+		const gridSize = Math.floor(
+			(isInVerticalOrientation ? width : height) / dataGroupNames.length
+		);
+		const boxWidth = Math.min(gridSize / 2, 16);
+		const boxStrokeWidth = '1';
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:57]: you don't want to put this in the non-customizable configurations?

File: packages/core/src/components/graphs/boxplot.ts [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+			)
+			.attr('stroke-width', boxStrokeWidth)
+			.attr('fill', 'none')
+			.transition(
+				this.services.transitions.getTransition(
+					'boxplot-update-verticalstartline',
+					animate
+				)
+			)
+
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:95]:
```suggestion
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:96]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+			)
+			.attr('stroke-width', boxStrokeWidth)
+			.attr('fill', 'none')
+			.transition(
+				this.services.transitions.getTransition(
+					'boxplot-update-verticalendline',
+					animate
+				)
+			)
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:126]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+				const y1 = cartesianScales.getRangeValue(d.quartiles.q_75);
+
+				return Tools.generateSVGPathString(
+					{ x0, x1, y0, y1 },
+					orientation
+				);
+			});
+
+		// The offset needed to center boxes on each group
+		const offset = gridSize / 2 - boxWidth / 2;
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:139]: doesnt seem like you use this?

File: packages/core/src/components/graphs/boxplot.ts [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+			.attr('stroke-width', boxStrokeWidth)
+			.attr('role', Roles.GRAPHICS_SYMBOL)
+			.attr('aria-roledescription', 'box')
+			.transition(
+				this.services.transitions.getTransition(
+					'boxplot-update-quartiles',
+					animate
+				)
+			)
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:166]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+
+		/*
+		 * Draw out and update highlight areas
+		 */
+		boxGroupsEnter
+			.append('path')
+			.merge(boxGroups.select('path.highlight-area'))
+			.attr('class', 'highlight-area')
+			.attr('opacity', 0)
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:198]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+			)
+			.attr('stroke-width', 2)
+			.attr('fill', 'none')
+			.transition(
+				this.services.transitions.getTransition(
+					'boxplot-update-startingwhisker',
+					animate
+				)
+			)
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:232]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+				})
+			)
+			.attr('stroke-width', 2)
+			.transition(
+				this.services.transitions.getTransition(
+					'boxplot-update-median',
+					animate
+				)
+			)
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:266]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+			)
+			.attr('stroke-width', 2)
+			.attr('fill', 'none')
+			.transition(
+				this.services.transitions.getTransition(
+					'boxplot-update-endingwhisker',
+					animate
+				)
+			)
+			.attr('d', (d, i) => {
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:300]:
```suggestion
			.attr('d', d => {
```

File: packages/core/src/components/graphs/boxplot.ts [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+					value: outlier,
+				};
+			})
+		);
+
+		circles.exit().remove();
+
+		const circlesEnter = circles.enter().append('circle');
+
+		const rangeIdentifier = this.services.cartesianScales.getRangeIdentifier();
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:332]: doesn't seem like you use this range id

File: packages/core/src/components/graphs/boxplot.ts [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+		circles
+			.merge(circlesEnter)
+			.attr('r', 4)
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:335]: non-customizable config ?

File: packages/core/src/components/graphs/boxplot.ts [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+			.attr('class', () =>
+				this.model.getColorClassName({
+					classNameTypes: [
+						ColorClassNameTypes.FILL,
+						ColorClassNameTypes.STROKE,
+					],
+					originalClassName: 'outlier',
+				})
+			)
+			.attr('fill-opacity', 0.3)
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:345]: also a config?

File: packages/core/src/components/graphs/boxplot.ts [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+
+		this.parent
+			.selectAll('path.highlight-area')
+			.on('mouseover', function (datum) {
+				const hoveredElement = select(this);
+				const parentElement = select(this.parentNode);
+				parentElement
+					.select('path.box')
+					.classed('hovered', true)
+					.attr('fill-opacity', 0.5);
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/boxplot.ts:373]: configuration for the opacity on hover?

**Review by natashadecoste — CHANGES_REQUESTED** [review]
the deploys for react & angular aren't working for me (on Chrome) 🤔 

angular:
image

react:
image

**Review by natashadecoste — APPROVED** [review]

**Review by sophiiae — APPROVED** [review]

**Pull request merged**